src/test-c-gen/function_return_builder.cpp

```diff
@@ -0,0 +1,203 @@
+ \*******************************************************************/
+function_return_buildert::function_return_buildert(const inputst &all_inputs,
+                                                   const irep_idt &function_id,
+                                                   expr2cleanct &e2c_converter)
+  :e2c(e2c_converter)
+{
+  has_return = false;
+
+  for(const input_entryt &input : all_inputs)
+  {
+    // TODO: this isn't a great way to find returns
+    if(input.first == "return'")
+    {
+      has_return = true;
+      return_entry = input;
+      set_return_variable_name(function_id);
+      return;
+    }
+  }
+
+  has_return = false;
+}
+
+/*******************************************************************\
+Function: function_return_buildert::set_return_variable_name
+Inputs:
+ function_id - The name of the function
+Purpose: Store the variable named used to store the result of the function
+ \*******************************************************************/
+void function_return_buildert::set_return_variable_name(
+    const irep_idt &function_id)
+{
+  std::ostringstream ret_name_builder;
+  ret_name_builder << "ret_";
+  ret_name_builder << function_id;
+  return_var_name = ret_name_builder.str();
+}
+
+/*******************************************************************\
+Function: function_return_buildert::add_assertions_for_expression
+Inputs:
+ correct_expression - The expression representing the value expected
+ ret_value_var - The name of the variable (including relevant nesting)
+Pupose: Add assertions to the assertions list for the return value
+ \*******************************************************************/
+void function_return_buildert::add_assertions_for_expression(const exprt &correct_expression,
+                                                             std::string ret_value_var)
+{
+  assert(has_return);
+
+  const typet &type = correct_expression.type();
+
+  if(type.id() == ID_struct)
+  {
+    add_assertions_for_struct_expression(correct_expression, ret_value_var);
+  }
+  else if(type.id() == ID_pointer)
+  {
+    // TODO - this should check pointers dereferenced == something sensible?
+    add_assertions_for_simple_expression(correct_expression, ret_value_var);
+  }
+  else
+  {
+    add_assertions_for_simple_expression(correct_expression, ret_value_var);
+  }
+}
+
+/*******************************************************************\
+Function: function_return_buildert::add_assertions_for_struct_expression
+Inputs:
+ correct_expression - The expression representing the value expected
+ ret_value_var - The name of the variable (including relevant nesting)
+Purpose: Look through all the components of the struct and add assertions for
+         each of them.
+ \*******************************************************************/
+void function_return_buildert::add_assertions_for_struct_expression(
+    const exprt &correct_expression, std::string ret_value_var)
```

Comment:
this will require some generalisation if we want to generate assertions for side effects too